rpcs3/xinput_pad_handler.cpp

```diff
@@ -2,7 +2,214 @@
 #ifdef _WIN32
 #include "xinput_pad_handler.h"
 
-xinput_pad_handler::xinput_pad_handler() : PadHandlerBase(pad_handler::xinput)
+#define NOMINMAX
+#include <Windows.h>
+#include <Xinput.h>
+
+class xinput_pad_processor_base
+{
+protected:
+	typedef DWORD(WINAPI* PFN_XINPUTSETSTATE)(DWORD, XINPUT_VIBRATION*);
+	typedef DWORD(WINAPI* PFN_XINPUTGETBATTERYINFORMATION)(DWORD, BYTE, XINPUT_BATTERY_INFORMATION*);
+
+protected:
+	bool is_init{false};
+	HMODULE library{nullptr};
+	PFN_XINPUTSETSTATE xinputSetState{nullptr};
+	PFN_XINPUTGETBATTERYINFORMATION xinputGetBatteryInformation{nullptr};
+
+public:
+	// This union should hold state structures of all implemented children of xinput_pad_processor_base
+	union XInputStateUnion
+	{
+		XINPUT_STATE xiState;
+	};
+
+public:
+	virtual ~xinput_pad_processor_base()
+	{
+		if (library)
+		{
+			FreeLibrary(library);
+		}
+	}
+
+	virtual pad_handler GetHandlerType() const = 0;
+	virtual const char* GetNameString() const = 0;
+	virtual const std::unordered_map<u32, std::string>& GetButtonList() const = 0;
+	virtual bool Init() = 0;
+	virtual DWORD GetState(DWORD userIndex, XInputStateUnion* state) = 0;
+	virtual std::array<u16, xinput_pad_handler::XInputKeyCodes::KeyCodeCount> GetButtonValues(const XInputStateUnion& state) = 0;
+
+	DWORD SetState(DWORD userIndex, u16 leftMotor, u16 rightMotor)
+	{
+		XINPUT_VIBRATION state { leftMotor, rightMotor };
+		return xinputSetState(userIndex, &state);
+	}
+
+	DWORD GetBatteryInformation(DWORD userIndex, BYTE devType, XINPUT_BATTERY_INFORMATION* batteryInformation)
+	{
+		return xinputGetBatteryInformation(userIndex, devType, batteryInformation);
+	}
+};
+
+class xinput_pad_processor_xi final : public xinput_pad_processor_base
+{
+	static constexpr DWORD GUIDE_BUTTON          = 0x0400;
+	static constexpr LPCWSTR LIBRARY_FILENAMES[] = {L"xinput1_4.dll", L"xinput1_3.dll", L"xinput9_1_0.dll"};
```

Comment:
what happened to 1_2

Reply:
also needs to solve conflicts

Reply:
I'll solve conflicts in a bit. XInput 1.2 is literally dead and superseded by XInput 1.3 - if user is running a machine which has 1.2 but not 1.3 nor 1.4, then it means it's an outdated Windows XP/Vista/7 which doesn't even have the latest DirectX End User Runtimes, thus probably doesn't have XAudio 2.7 either, and possibly more. I can add it back if really needed, but this is kind of like supporting something which is dead and replaced by a newer fixed solution.

Reply:
Also, even on those "broken systems" I described earlier xinput9_1_0.dll is **guaranteed** to exist (it's shipped with any Windows starting with XP), so I would consider it a better fallback than a DLL nobody remembers about, MSDN docs included.